sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_pointcloud_preprocessor/cuda_pointcloud_preprocessor.cu

```diff
@@ -0,0 +1,692 @@
+
+  std::size_t twist_index = 0;
+  std::size_t angular_velocity_index = 0;
+
+  for (; twist_index < twist_queue.size() ||
+         angular_velocity_index < angular_velocity_queue.size();) {
+    uint64_t twist_stamp, input_twist_global_stamp_nsec, angular_velocity_global_stamp_nsec;
+
+    if (twist_index < twist_queue.size()) {
+      input_twist_global_stamp_nsec =
+        1'000'000'000 * static_cast<uint64_t>(twist_queue[twist_index].header.stamp.sec) +
+        static_cast<uint64_t>(twist_queue[twist_index].header.stamp.nanosec);
+      v.x() = twist_queue[twist_index].twist.twist.linear.x;
+      v.y() = twist_queue[twist_index].twist.twist.linear.y;
+      v.z() = twist_queue[twist_index].twist.twist.linear.z;
+    } else {
+      input_twist_global_stamp_nsec = std::numeric_limits<uint64_t>::max();
+      v = Eigen::Vector3f::Zero();
+    }
+
+    if (angular_velocity_index < angular_velocity_queue.size()) {
+      angular_velocity_global_stamp_nsec =
+        1'000'000'000 *
+          static_cast<uint64_t>(angular_velocity_queue[angular_velocity_index].header.stamp.sec) +
+        static_cast<uint64_t>(angular_velocity_queue[angular_velocity_index].header.stamp.nanosec);
+      w.x() = angular_velocity_queue[angular_velocity_index].vector.x;
+      w.y() = angular_velocity_queue[angular_velocity_index].vector.y;
+      w.z() = angular_velocity_queue[angular_velocity_index].vector.z;
+    } else {
+      angular_velocity_global_stamp_nsec = std::numeric_limits<uint64_t>::max();
+      w = Eigen::Vector3f::Zero();
+    }
+
+    if (input_twist_global_stamp_nsec < angular_velocity_global_stamp_nsec) {
+      twist_stamp = input_twist_global_stamp_nsec;
+      twist_index++;
+    } else if (input_twist_global_stamp_nsec > angular_velocity_global_stamp_nsec) {
+      twist_stamp = angular_velocity_global_stamp_nsec;
+      angular_velocity_index++;
+    } else {
+      twist_index++;
+      angular_velocity_index++;
+    }
+
+    TwistStruct3D twist;
+
+    Eigen::Map<Eigen::Matrix4f> cum_transform_buffer_map(twist.cum_transform_buffer);
+    Eigen::Map<Eigen::Vector3f> v_map(twist.v);
+    Eigen::Map<Eigen::Vector3f> w_map(twist.w);
+    cum_transform_buffer_map = cum_transform;
+
+    uint64_t twist_global_stamp_nsec = twist_stamp;
+    assert(twist_global_stamp_nsec > pointcloud_stamp_nsec);  // by construction
+    uint32_t twist_from_pointcloud_start_nsec = twist_global_stamp_nsec - pointcloud_stamp_nsec;
+
+    twist.stamp_nsec = twist_from_pointcloud_start_nsec;
+    v_map = v;
+    w_map = w;
+    twist.last_stamp_nsec = last_stamp_nsec;
+    host_twist_structs.push_back(twist);
+
+    double dt_seconds = 1e-9 * (twist.stamp_nsec - last_stamp_nsec);
+    last_stamp_nsec = twist.stamp_nsec;
+
+    auto delta_transform = transformationMatrixFromVelocity(v, w, dt_seconds);
+    cum_transform = cum_transform * delta_transform;
+  }
+
+  // Copy to device
+  device_twist_3d_structs_ = host_twist_structs;
```

Comment:
I suspect this code may cause GPU memory fragmentation and/or repeated implicit GPU memory allocation, both of which harm performance. To prevent such kind of issues, I would suggest using CUDA's memory pool functionality. I'll add suggestion code to the constructor of `CudaPointcloudPreprocessor`.